### ISSUE TITLE
[BUG] Unresponsive minions when master hostname not resolvable in DNS

### ISSUE THREAD

**Issue by gbunt** [saltstack/salt#65139]
**Description**
Local master hostname is not in DNS, though registered in `/etc/hosts`. Yet, master is still performing continuous lookups, causing minion responses to be very slow, or simply timing out.
This was initially run into when the configured DNS server was timing out, but after having addressed that, the issue still persists. Note that this specifically happens in this EC2 environment, don't seem to be able to reproduce this on a local test environment using the same versions salt and OS.

**Setup**
- master config:
```
interface: 192.168.2.11
user: root
timeout: 30
max_event_size: 10485760
extension_modules: /opt/corp-salt/8.0/cluster/modules
presence_events: True
worker_threads: 6
file_roots:
  version-8.0:
    - /opt/corp-salt/8.0/cluster/states
    - /opt/corp-salt/8.0/cluster/formulas
    - /opt/corp-salt/8.0/cluster/modules
    - /opt/corp-salt/8.0/cluster/orch
  base:
    - /opt/corp-salt/7.5.2/cluster/states
    - /opt/corp-salt/7.5.2/cluster/formulas
    - /opt/corp-salt/7.5.2/cluster/modules
    - /opt/corp-salt/7.5.2/cluster/orch
top_file_merging_strategy: same
pillar_roots:
  version-8.0:
    - /opt/corp-salt/8.0/cluster/pillar
  base:
    - /opt/corp-salt/7.5.2/cluster/pillar
ext_pillar:
  - stack:
      opts:saltenv:
        version-8.0:
          - /opt/corp-salt/8.0/cluster/pillar/corp_base/pillarstack.cfg
          - /opt/corp-salt/8.0/cluster/pillar/corp_base/classstack.cfg
reactor:
  - 'salt/beacon/*/service/corp-*':
    - /opt/corp-salt/8.0/cluster/reactor/sync_dns.sls
  - 'salt/presence/change':
    - /opt/corp-salt/8.0/cluster/reactor/sync_dns.sls
reactor_worker_threads: 128
reactor_worker_hwm: 0
log_level: warning
tcp_keepalive: True
tcp_keepalive_idle: 10
tcp_keepalive_cnt: 3
tcp_keepalive_intvl: 10
```

- minion config:
```
master: 192.168.2.11
master_tries: -1
mine_enabled: True
mine_interval: 1
log_level: warning
saltenv: version-8.0
lock_saltenv: True
state_top_saltenv: version-8.0
pillarenv_from_saltenv: True
systemd.scope: False
```

Please be as specific as possible and give set-up details.

- [ ] on-prem machine
- [ ] VM (Virtualbox, KVM, etc. please specify)
- [x] VM running on a cloud service, please be explicit and add details
- [ ] container (Kubernetes, Docker, containerd, etc. please specify)
- [ ] or a combination, please be explicit
- [ ] jails if it is FreeBSD
- [ ] classic packaging
- [ ] onedir packaging
- [ ] used bootstrap to install

> EC2 Nitro instance running CentOS Linux release 7.8.2003 (Core) image

**Steps to Reproduce the behavior**
- Env is running a resolver locally:
```
# /etc/resolv.conf
; generated by /usr/sbin/dhclient-script
search ec2.internal
nameserver 127.0.0.1
```
- Host is in `/etc/hosts`:
```
192.168.1.11    n1.ec2.internal n1
```
- nsswitch orders `files` first:
```
# /etc/nsswitch.conf
hosts:      files dns myhostname
```
- Resolving fails instantly (expected/good):
```
$ host n1
Host n1 not found: 3(NXDOMAIN)
```
- `socket.getaddrinfo` returns instantly:
```
$ /opt/saltstack/salt/bin/python3
Python 3.10.11 (main, May  5 2023, 02:31:54) [GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import socket
>>> socket.getaddrinfo("n1", 0)
[(<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('192.168.1.11', 0)), (<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_DGRAM: 2>, 17, '', ('192.168.1.11', 0)), (<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_RAW: 3>, 0, '', ('192.168.1.11', 0)), (<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('192.168.1.11', 0)), (<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_DGRAM: 2>, 17, '', ('192.168.1.11', 0)), (<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_RAW: 3>, 0, '', ('192.168.1.11', 0))]
```
- Latency is low (tested longer window):
```
# to remote minion
64 bytes from n12.ec2.internal (192.168.1.25): icmp_seq=1 ttl=64 time=0.369 ms
# from remote minion to master
64 bytes from n1.ec2.internal (192.168.1.11): icmp_seq=1 ttl=64 time=0.162 ms
```
- `test.ping`, if it even responds, is very slow:
```
$ time salt-call test.ping
local:
    True

real	0m17.223s
```
- `test.ping` targeting all minions while running a packet trace on port 53, is instant, then seeing lots of lookups, then the remaining minions fail:
```
$ [root@n1 orch]# salt '*' test.ping
997e7460-83fa-44e9-a9d0-2a95a4ff2fe1:
    True
544a0d77-a203-4e60-9c1f-4e5c1ec7516a:
    True
c6fecfaa-ded1-445a-800e-46e18a07f778:
    True
84147b4b-85e2-4c7a-bf75-2d3e506eaf72:
    True
0a098a0e-36f8-4d2f-8b16-70ac4412459f:
    True
5aad7de2-94d5-4889-9b5f-581f018ea32c:
    True
22e12d2d-cbcc-4dc7-9579-e1e1dbef78a5:
    True
10:54:55.389293 IP 127.0.0.1.34512 > 127.0.0.1.domain: 62501+ AAAA? n1.ec2.internal. (33)
10:54:55.389391 IP 127.0.0.1.domain > 127.0.0.1.34512: 62501 NXDomain 0/0/0 (33)
10:54:55.389440 IP 127.0.0.1.52367 > 127.0.0.1.domain: 42526+ AAAA? n1. (20)
10:54:55.389478 IP 127.0.0.1.domain > 127.0.0.1.52367: 42526 NXDomain 0/0/0 (20)
10:54:55.862263 IP 127.0.0.1.33550 > 127.0.0.1.domain: 40168+ AAAA? n1.ec2.internal. (33)
10:54:55.862621 IP 127.0.0.1.domain > 127.0.0.1.33550: 40168 NXDomain 0/0/0 (33)
10:54:55.862718 IP 127.0.0.1.44090 > 127.0.0.1.domain: 30777+ AAAA? n1. (20)
10:54:55.862779 IP 127.0.0.1.domain > 127.0.0.1.44090: 30777 NXDomain 0/0/0 (20)
10:54:56.353702 IP 127.0.0.1.44592 > 127.0.0.1.domain: 64526+ AAAA? n1.ec2.internal. (33)
10:54:56.354056 IP 127.0.0.1.domain > 127.0.0.1.44592: 64526 NXDomain 0/0/0 (33)
10:54:56.354336 IP 127.0.0.1.56103 > 127.0.0.1.domain: 34905+ AAAA? n1. (20)
10:54:56.354537 IP 127.0.0.1.domain > 127.0.0.1.56103: 34905 NXDomain 0/0/0 (20)
10:54:56.814463 IP 127.0.0.1.50317 > 127.0.0.1.domain: 38538+ AAAA? n1.ec2.internal. (33)
10:54:56.814592 IP 127.0.0.1.domain > 127.0.0.1.50317: 38538 NXDomain 0/0/0 (33)
10:54:56.814699 IP 127.0.0.1.38805 > 127.0.0.1.domain: 17939+ AAAA? n1. (20)
10:54:56.814820 IP 127.0.0.1.domain > 127.0.0.1.38805: 17939 NXDomain 0/0/0 (20)
10:54:57.231747 IP 127.0.0.1.40933 > 127.0.0.1.domain: 626+ AAAA? n1.ec2.internal. (33)
10:54:57.231831 IP 127.0.0.1.domain > 127.0.0.1.40933: 626 NXDomain 0/0/0 (33)
10:54:57.231937 IP 127.0.0.1.55338 > 127.0.0.1.domain: 24273+ AAAA? n1. (20)
10:54:57.231979 IP 127.0.0.1.domain > 127.0.0.1.55338: 24273 NXDomain 0/0/0 (20)
10:54:57.587541 IP 127.0.0.1.38368 > 127.0.0.1.domain: 28895+ AAAA? n1.ec2.internal. (33)
10:54:57.587624 IP 127.0.0.1.domain > 127.0.0.1.38368: 28895 NXDomain 0/0/0 (33)
10:54:57.587663 IP 127.0.0.1.46204 > 127.0.0.1.domain: 29719+ AAAA? n1. (20)
10:54:57.587694 IP 127.0.0.1.domain > 127.0.0.1.46204: 29719 NXDomain 0/0/0 (20)
10:54:57.980921 IP 127.0.0.1.55802 > 127.0.0.1.domain: 64747+ AAAA? n1.ec2.internal. (33)
10:54:57.981035 IP 127.0.0.1.domain > 127.0.0.1.55802: 64747 NXDomain 0/0/0 (33)
10:54:57.981075 IP 127.0.0.1.38604 > 127.0.0.1.domain: 39546+ AAAA? n1. (20)
10:54:57.981108 IP 127.0.0.1.domain > 127.0.0.1.38604: 39546 NXDomain 0/0/0 (20)
10:55:03.870245 IP 127.0.0.1.47009 > 127.0.0.1.domain: 4195+ AAAA? n1.ec2.internal. (33)
10:55:03.870511 IP 127.0.0.1.domain > 127.0.0.1.47009: 4195 NXDomain 0/0/0 (33)
10:55:03.870773 IP 127.0.0.1.42039 > 127.0.0.1.domain: 52907+ AAAA? n1. (20)
10:55:03.870822 IP 127.0.0.1.domain > 127.0.0.1.42039: 52907 NXDomain 0/0/0 (20)
10:55:04.310897 IP 127.0.0.1.42550 > 127.0.0.1.domain: 54851+ AAAA? n1.ec2.internal. (33)
10:55:04.310988 IP 127.0.0.1.domain > 127.0.0.1.42550: 54851 NXDomain 0/0/0 (33)
10:55:04.311025 IP 127.0.0.1.34263 > 127.0.0.1.domain: 19978+ AAAA? n1. (20)
10:55:04.311054 IP 127.0.0.1.domain > 127.0.0.1.34263: 19978 NXDomain 0/0/0 (20)
10:55:05.098917 IP 127.0.0.1.45227 > 127.0.0.1.domain: 36553+ AAAA? n1.ec2.internal. (33)
10:55:05.099217 IP 127.0.0.1.domain > 127.0.0.1.45227: 36553 NXDomain 0/0/0 (33)
10:55:05.099381 IP 127.0.0.1.37798 > 127.0.0.1.domain: 46298+ AAAA? n1. (20)
10:55:05.099464 IP 127.0.0.1.domain > 127.0.0.1.37798: 46298 NXDomain 0/0/0 (20)
10:55:05.629670 IP 127.0.0.1.56562 > 127.0.0.1.domain: 3136+ AAAA? n1.ec2.internal. (33)
10:55:05.629815 IP 127.0.0.1.domain > 127.0.0.1.56562: 3136 NXDomain 0/0/0 (33)
10:55:05.629862 IP 127.0.0.1.40763 > 127.0.0.1.domain: 12025+ AAAA? n1. (20)
10:55:05.629953 IP 127.0.0.1.domain > 127.0.0.1.40763: 12025 NXDomain 0/0/0 (20)
10:55:06.222723 IP 127.0.0.1.37130 > 127.0.0.1.domain: 63087+ AAAA? n1.ec2.internal. (33)
10:55:06.222820 IP 127.0.0.1.domain > 127.0.0.1.37130: 63087 NXDomain 0/0/0 (33)
10:55:06.222874 IP 127.0.0.1.58962 > 127.0.0.1.domain: 49307+ AAAA? n1. (20)
10:55:06.222973 IP 127.0.0.1.domain > 127.0.0.1.58962: 49307 NXDomain 0/0/0 (20)
10:55:06.618507 IP 127.0.0.1.53455 > 127.0.0.1.domain: 33009+ AAAA? n1.ec2.internal. (33)
10:55:06.618606 IP 127.0.0.1.domain > 127.0.0.1.53455: 33009 NXDomain 0/0/0 (33)
10:55:06.618759 IP 127.0.0.1.45343 > 127.0.0.1.domain: 6892+ AAAA? n1. (20)
10:55:06.619753 IP 127.0.0.1.domain > 127.0.0.1.45343: 6892 NXDomain 0/0/0 (20)
10:55:07.075149 IP 127.0.0.1.59153 > 127.0.0.1.domain: 12944+ AAAA? n1.ec2.internal. (33)
10:55:07.075246 IP 127.0.0.1.domain > 127.0.0.1.59153: 12944 NXDomain 0/0/0 (33)
10:55:07.075300 IP 127.0.0.1.48489 > 127.0.0.1.domain: 23172+ AAAA? n1. (20)
10:55:07.075374 IP 127.0.0.1.domain > 127.0.0.1.48489: 23172 NXDomain 0/0/0 (20)
f1979ee3-83b7-4870-bdda-78e8513a77a0:
    Minion did not return. [Not connected]
c88a1031-65ae-4903-9d35-098cebda6817:
    Minion did not return. [No response]
    The minions may not have all finished running and any remaining minions will return upon completion. To look up the return data for this job later, run the following command:
    
    salt-run jobs.lookup_jid 20230907105437440849
59832408-76e5-4eeb-a162-98a2909b94d5:
    Minion did not return. [No response]
    The minions may not have all finished running and any remaining minions will return upon completion. To look up the return data for this job later, run the following command:
    
    salt-run jobs.lookup_jid 20230907105437440849
5ba9f2fb-0c3c-45d6-a30a-147204cc9d97:
    Minion did not return. [Not connected]
90267d7e-7694-41f2-9725-4a322c4ef6f7:
    Minion did not return. [No response]
    The minions may not have all finished running and any remaining minions will return upon completion. To look up the return data for this job later, run the following command:
    
    salt-run jobs.lookup_jid 20230907105437440849
b1781d62-e145-4475-98ca-123ccea5ea19:
    Minion did not return. [No response]
    The minions may not have all finished running and any remaining minions will return upon completion. To look up the return data for this job later, run the following command:
    
    salt-run jobs.lookup_jid 20230907105437440849
7e94ce4d-80e5-4c1f-80f0-cb478381b5b9:
    Minion did not return. [Not connected]
ERROR: Minions returned with non-zero exit code
```
On every attempt, different minions fail. This obviously isn't just affecting the test module, but pretty much renders all of salt broken.

**Expected behavior**
Master doesn't perform DNS lookups if master hostname is registered in `/etc/hosts` and nsswitch orders `files` first. Even if it would do a DNS lookup and that would fail instantly, the master shouldn't lock up.

**Screenshots**

**Versions Report**
<details><summary>salt --versions-report</summary>
(Provided by running salt --versions-report. Please also mention any differences in master/minion versions.)

```yaml
Salt Version:
          Salt: 3006.1
 
Python Version:
        Python: 3.10.11 (main, May  5 2023, 02:31:54) [GCC 11.2.0]
 
Dependency Versions:
          cffi: 1.14.6
      cherrypy: unknown
      dateutil: 2.8.1
     docker-py: Not Installed
         gitdb: Not Installed
     gitpython: Not Installed
        Jinja2: 3.1.2
       libgit2: Not Installed
  looseversion: 1.0.2
      M2Crypto: Not Installed
          Mako: Not Installed
       msgpack: 1.0.2
  msgpack-pure: Not Installed
  mysql-python: Not Installed
     packaging: 22.0
     pycparser: 2.21
      pycrypto: Not Installed
  pycryptodome: 3.9.8
        pygit2: Not Installed
  python-gnupg: 0.4.8
        PyYAML: 5.4.1
         PyZMQ: 23.2.0
        relenv: 0.12.3
         smmap: Not Installed
       timelib: 0.2.4
       Tornado: 4.5.3
           ZMQ: 4.3.4
 
System Versions:
          dist: centos 7.8.2003 Core
        locale: utf-8
       machine: x86_64
       release: 3.10.0-1127.el7.x86_64
        system: Linux
       version: CentOS Linux 7.8.2003 Core
```
</details>

**Additional context**

**Comment by OrangeDog** [comment]
Those are AAAA requests, i.e. IPv6, but all your testing is with IPv4.

**Comment by gbunt** [comment]
Thanks, not sure how i didn't notice that!

However, that may have been a red herring. Partially anyway.

After adding an ipv6 entry in `/etc/hosts` (note that the hosts do not have an ipv6 address attached to any interface though) things seems to have somewhat stabilized for the v3006.1 minions, but not the v2019.2.5 ones (this is an upgrade scenario. Yes, should've mentioned that).
Have also disabled presence events for now, without any noticeable effect. A `'*' test.ping` still times out on arbitrary minions each run while running various connectivity checks on network instability has been found (and this env also hasn't been running into any issues whatsoever with a v2019 master and all v2019 minions).

We run a `test.ping` healthcheck before an orchestrated (not salt orch) upgrade of salt. Considering to disable that, see what happens.

**Comment by OrangeDog** [comment]
2019.2.5 is not supported I'm afraid (and 2019.2.8 is the newest security patch on that line).

**Comment by gbunt** [comment]
@OrangeDog no idea whether 2019.2.8 would've helped anything, the issue was really specific to that EC2 env somehow. No other environments we've tested upgrade path 2019.2.5 weren't affected.
For our specific use-case and this environment, we've now simply skipped running healthchecks in our upgrade orchestration after the master was upgraded, but the minions weren't yet. The issue was gone when all minions were upgraded.

Let's close this issue. Thanks.
